### PR TITLE
[query] Add Prometheus engine metrics and upgrade to Prometheus 2.17.2

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

This adds metrics for the Prometheus engine combined with normal metrics exposed by the coordinator.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -33,6 +33,7 @@ thrift_output_dir    := generated/thrift/rpc
 thrift_rules_dir     := generated/thrift
 vendor_prefix        := vendor
 bad_trace_dep        := go.etcd.io/etcd/vendor/golang.org/x/net/trace
+bad_prom_vendor_dir  := github.com/prometheus/prometheus/vendor
```
Comment (arnikola) [inline Makefile:36]: :(

File: src/x/instrument/config_prometheus.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+	prom "github.com/m3db/prometheus_client_golang/prometheus"
+	"github.com/m3db/prometheus_client_golang/prometheus/promhttp"
+	dto "github.com/m3db/prometheus_client_model/go"
+	extprom "github.com/prometheus/client_golang/prometheus"
+	"github.com/uber-go/tally/prometheus"
+)
+
+const (
+	promNameLabel = "__name__"
```
Comment (arnikola) [inline src/x/instrument/config_prometheus.go:38]: This should probably take a models.TagOptions and extract name from that

File: src/x/instrument/config_prometheus.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+			}
+		}
+	}
+
+	switch c.TimerType {
+	case "summary":
+		opts.DefaultTimerType = prometheus.SummaryTimerType
+	case "histogram":
+		opts.DefaultTimerType = prometheus.HistogramTimerType
+	}
```
Comment (arnikola) [inline src/x/instrument/config_prometheus.go:144]: Should this have a default case?

File: src/x/instrument/config_prometheus.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+	if len(c.DefaultHistogramBuckets) > 0 {
+		var values []float64
```
Comment (arnikola) [inline src/x/instrument/config_prometheus.go:147]: nit; values := make([]float64, 0, len(c.DefaultHIstogramBuckets))

File: src/x/instrument/config_prometheus.go [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+		for _, value := range c.DefaultHistogramBuckets {
+			values = append(values, value.Upper)
+		}
+		opts.DefaultHistogramBuckets = values
+	}
+
+	if len(c.DefaultSummaryObjectives) > 0 {
+		values := make(map[float64]float64)
```
Comment (arnikola) [inline src/x/instrument/config_prometheus.go:155]: nit; `values := make(map[float64]float64, len(c.DefaultHIstogramBuckets))`

File: src/x/instrument/config_prometheus.go [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+		}
+
+		for _, elem := range gathered {
+			entry := &dto.MetricFamily{
+				Name:   elem.Name,
+				Help:   elem.Help,
+				Metric: make([]*dto.Metric, 0, len(elem.Metric)),
+			}
+
+			if v := elem.Type; v != nil {
```
Comment (arnikola) [inline src/x/instrument/config_prometheus.go:230]: should ` v == nil` fail here?

File: src/x/instrument/config_prometheus.go [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+			}
+
+			for _, metricElem := range elem.Metric {
+				metricEntry := &dto.Metric{
+					Label:       make([]*dto.LabelPair, 0, len(metricElem.Label)),
+					TimestampMs: metricElem.TimestampMs,
+				}
+
+				if v := metricElem.Gauge; v != nil {
+					metricEntry.Gauge = &dto.Gauge{
```
Comment (arnikola) [inline src/x/instrument/config_prometheus.go:242]: Can different metrics have multiple metric types or should these continue?

File: src/x/instrument/config_prometheus.go [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+				for _, labelElem := range metricElem.Label {
+					var labelName *string
+					if labelElem.Name != nil {
+						labelName = labelElem.Name
+					}
+
+					// Allow prefixing metric names.
+					if secondary.SubScope != "" &&
+						labelName != nil &&
+						*labelName == promNameLabel {
```
Comment (arnikola) [inline src/x/instrument/config_prometheus.go:302]: nit: use tag options here for the name label

File: src/x/instrument/config_prometheus.go (hunk 2 above)
Comment (robskillington) [reply src/x/instrument/config_prometheus.go:38]: I'd agree, although this is very specifically for export of prometheus metrics (using prometheus metric libraries).

I'd rather not make the `src/x/instrument` package rely on `src/query/models` if that's ok?

File: src/x/instrument/config_prometheus.go (hunk 6 above)
Comment (robskillington) [reply src/x/instrument/config_prometheus.go:230]: I'm really just exporting whatever the exporter is already gathering, it's pure pass through here - think of it as a serialization issue (lol).

File: src/x/instrument/config_prometheus.go (hunk 7 above)
Comment (robskillington) [reply src/x/instrument/config_prometheus.go:242]: I'm really just exporting whatever the exporter is already gathering, it's pure pass through here - think of it as a serialization issue (lol).

File: Makefile (hunk 1 above)
Comment (robskillington) [reply Makefile:36]: Yeah =[

File: src/x/instrument/config_prometheus.go (hunk 3 above)
Comment (robskillington) [reply src/x/instrument/config_prometheus.go:143]: The default value for `DefaultTimerType` is actually valid hence why it is not handled here.

**Pull request merged**